### PR TITLE
Fix search_payments method in ActiveMerchant::PaymentPlugin

### PR DESCRIPTION
**Pull request by larrylv**
Related issue: https://github.com/killbill/killbill-java-parser/pull/9

With my patch in https://github.com/killbill/killbill-java-parser/pull/9, this spec should pass successfully.

```ruby
module Killbill
  module Plugin
    module Model
      class EnumeratorIterator
        alias_method :hasNext, :has_next
      end
    end
  end
end
```

Without the patch, it will raise an error:

``` ruby
 1) Killbill::Plugin::ActiveMerchant::PaymentPlugin when skipping the gateway should support search_payments                                                      [8/46]
     Failure/Error: results = plugin.search_payments(@kb_payment_id, 0, 100, [], @call_context).to_a
     NoMethodError:
       undefined method `hasNext' for #<Killbill::Plugin::Model::EnumeratorIterator:0x3663d0ba>
     # file:/Users/Larry/.rbenv/versions/jruby-1.7.20/lib/jruby.jar!/jruby/java/java_ext/java.lang.rb:12:in `each'
     # ./spec/killbill/helpers/payment_plugin_spec.rb:318:in `(root)'
```

Also, the iterator should just be an iterator, instead of converting it in `to_java` method, which will not work in ruby codes.

### CODE REVIEW

File: lib/killbill/helpers/active_merchant/payment_plugin.rb [hunk 1]
```diff
@@ -146,7 +146,7 @@ def get_payment_info(kb_account_id, kb_payment_id, properties, context)
 
         def search_payments(search_key, offset = 0, limit = 100, properties, context)
           options = properties_to_hash(properties)
-          @response_model.search(search_key, context.tenant_id, offset, limit, :payment)
+          @response_model.search(search_key, context.tenant_id, offset, limit)
```
Comment (pierre) [inline lib/killbill/helpers/active_merchant/payment_plugin.rb]: :+1: That's a bug indeed (see 42fe22bd41c47d267447eed4d43776337f2c9dcd).

File: spec/killbill/helpers/payment_plugin_spec.rb [hunk 2]
```diff
@@ -308,6 +308,17 @@
       source.first_name.should == 'John'
       source.last_name.should == 'Doe'
     end
+
+    it "should support search_payments" do
+      response = ::Killbill::Test::TestResponse.create(
+        api_call: :purchase, kb_payment_id: @kb_payment_id,
```
Comment (pierre) [inline spec/killbill/helpers/payment_plugin_spec.rb]: FYI, we want to stick to the old hash syntax for now.

File: lib/killbill/gen/api/pagination.rb [hunk 3]
```diff
@@ -41,7 +41,7 @@ def initialize()
 
         def to_java()
           # conversion for iterator [type = java.util.Iterator]
-          @iterator = Killbill::Plugin::Model::EnumeratorIterator.new(@iterator)
```
Comment (pierre) [inline lib/killbill/gen/api/pagination.rb]: Why this change?

Comment (larrylv) [inline lib/killbill/gen/api/pagination.rb]: You could see my change [here](https://github.com/killbill/killbill-plugin-framework-ruby/pull/50/files#diff-632ff814489142f5b67ec628c1fc0ec3R205). The iterator is already a `Killbill::Plugin::Model::EnumeratorIterator instance, so I think there is no need to do that?

Please correct me if they are irrelevant.

File: spec/killbill/helpers/payment_plugin_spec.rb (hunk 2 above)
Comment (larrylv) [inline spec/killbill/helpers/payment_plugin_spec.rb]: Ok, updated to old hash syntax and squashed my commit.

File: lib/killbill/gen/api/pagination.rb (hunk 3 above)
Comment (larrylv) [inline lib/killbill/gen/api/pagination.rb]: Hi @pierre , any updates? Please inform me if this change is not necessary, I will update my pr. Thanks.

Comment (pierre) [inline lib/killbill/gen/api/pagination.rb]: > Hi @pierre , any updates?

Sorry for the slow reply, but I was :airplane:  in the past two weeks.

>  Please inform me if this change is not necessary, I will update my pr. Thanks.

In general, you should never do any manual change in the `gen` package as the code is auto-generated by the [killbill-java-parser](https://github.com/killbill/killbill-java-parser).

But regardless, this change seems incorrect. `EnumeratorIterator` is a class to convert a Ruby `Enumerable` or `Enumerator` to a `java.util.Iterator`. See [here](https://github.com/killbill/killbill-java-parser/blob/1967e802bbb2bbf2ae4a193832e90d0941c46ad3/src/main/java/com/ning/killbill/generators/ruby/JRubyPluginGenerator.java#L659). It should only be used in the `to_java` API conversions behind the scenes (the core of the `killbill` gem is pure Ruby, not JRuby).

File: spec/killbill/helpers/payment_plugin_spec.rb [hunk 4]
```diff
@@ -308,6 +308,17 @@
...
       source.last_name.should == 'Doe'
     end
+
+    it "should support search_payments" do
+      response = ::Killbill::Test::TestResponse.create(
+        :api_call => :purchase, :kb_payment_id => @kb_payment_id,
+        :kb_account_id => @kb_account_id, :kb_tenant_id => @call_context.tenant_id,
+        :success => true, :created_at => Time.now, :updated_at => Time.now
+      )
+      results = plugin.search_payments(@kb_payment_id, 0, 100, [], @call_context).to_a
```
Comment (pierre) [inline spec/killbill/helpers/payment_plugin_spec.rb]: The test doesn't look right. The [search_payments](https://github.com/killbill/killbill-plugin-api/blob/master/payment/src/main/java/org/killbill/billing/payment/plugin/api/PaymentPluginApi.java#L159) API returns a [Pagination](https://github.com/killbill/killbill-api/blob/master/src/main/java/org/killbill/billing/util/entity/Pagination.java) object, which doesn't have a `to_a` method (I'm guessing JRuby injects a `to_a` for convenience, provided `hasNext` and `next` are implemented).

The test should pass if you replace `to_a` with `.iterator.to_a`.

That being said, going back to the original [bug](https://github.com/killbill/killbill-java-parser/pull/9), did you manage to trigger the issue by calling the API from Kill Bill, or via a similar test (I see `./spec/ach_chase/base_plugin_spec.rb` in the backtrace)?

Comment (larrylv) [inline spec/killbill/helpers/payment_plugin_spec.rb]: @pierre Thanks for your time.

I did trigger the issue by a similar test instead of calling API from Kill Bill. Sorry for my incorrect usage.

I will modify my pr to fix `@response_model.search(search_key, context.tenant_id, offset, limit, :payment)` -> `@response_model.search(search_key, context.tenant_id, offset, limit)`

File: lib/killbill/gen/api/pagination.rb (hunk 3 above)
Comment (larrylv) [inline lib/killbill/gen/api/pagination.rb]: @pierre Thanks for you explanation.

**Comment by pierre** [comment]
:tada:
